piker/brokers/binance.py

```diff
@@ -0,0 +1,496 @@
+
+"""
+from contextlib import asynccontextmanager, AsyncExitStack
+from types import ModuleType
+from typing import List, Dict, Any, Tuple, Union, Optional
+import json
+import time
+
+import trio_websocket
+from trio_typing import TaskStatus
+from trio_websocket._impl import (
+    ConnectionClosed,
+    DisconnectionTimeout,
+    ConnectionRejected,
+    HandshakeError,
+    ConnectionTimeout,
+)
+
+import arrow
+import asks
+import numpy as np
+import trio
+import tractor
+from pydantic.dataclasses import dataclass
+from pydantic import BaseModel
+
+
+from .api import open_cached_client
+from ._util import resproc, SymbolNotFound
+from ..log import get_logger, get_console_log
+from ..data import ShmArray
+
+log = get_logger(__name__)
+
+
+_url = 'https://api.binance.com'
+
+
+# Broker specific ohlc schema (rest)
+_ohlc_dtype = [
+    ('index', int),
+    ('time', int),
+    ('open', float),
+    ('high', float),
+    ('low', float),
+    ('close', float),
+    ('volume', float),
+    ('bar_wap', float),  # will be zeroed by sampler if not filled
+
+    # XXX: some additional fields are defined in the docs:
+    # https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-data
+
+    # ('close_time', int),
+    # ('quote_vol', float),
+    # ('num_trades', int),
+    # ('buy_base_vol', float),
+    # ('buy_quote_vol', float),
+    # ('ignore', float),
+]
+
+# UI components allow this to be declared such that additional
+# (historical) fields can be exposed.
+ohlc_dtype = np.dtype(_ohlc_dtype)
+
+_show_wap_in_history = False
+
+
+# https://binance-docs.github.io/apidocs/spot/en/#exchange-information
+class Pair(BaseModel):
+    symbol: str
+    status: str
+
+    baseAsset: str
+    baseAssetPrecision: int
+    quoteAsset: str
+    quotePrecision: int
+    quoteAssetPrecision: int
+
+    baseCommissionPrecision: int
+    quoteCommissionPrecision: int
+
+    orderTypes: List[str]
+
+    icebergAllowed: bool
+    ocoAllowed: bool
+    quoteOrderQtyMarketAllowed: bool
+    isSpotTradingAllowed: bool
+    isMarginTradingAllowed: bool
+
+    filters: List[Dict[str, Union[str, int, float]]]
+    permissions: List[str]
+
+
+@dataclass
+class OHLC:
+    """Description of the flattened OHLC quote format.
+
+    For schema details see:
+    https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-streams
+
+    """
+    time: int
+
+    open: float
+    high: float
+    low: float
+    close: float
+    volume: float
+
+    close_time: int
+
+    quote_vol: float
+    num_trades: int
+    buy_base_vol: float
+    buy_quote_vol: float
+    ignore: int
+
+    # null the place holder for `bar_wap` until we
+    # figure out what to extract for this.
+    bar_wap: float = 0.0
+
+
+# convert arrow timestamp to unixtime in miliseconds
+def binance_timestamp(when):
+    return int((when.timestamp() * 1000) + (when.microsecond / 1000))
+
+
+class Client:
+
+    def __init__(self) -> None:
+        self._sesh = asks.Session(connections=4)
```

Comment:
Oh that was the other thing.

What do you think about swapping out [`httpx`](https://www.python-httpx.org/async/) here?

Apparently everyone in `trio` is saying it's the more maintained project?

Reply:
We can defer to #135 as well if you want.